src/main/java/uk/ac/ebi/ampt2d/commons/accession/persistence/BasicSpringDataRepositoryDatabaseService.java

```diff
@@ -37,105 +42,189 @@
+                .collect(Collectors.toList());
+    }
+
+    private ACCESSION_ENTITY filterOldVersions(List<ACCESSION_ENTITY> accessionedElements) {
+        int maxVersion = 1;
+        ACCESSION_ENTITY lastVersionEntity = null;
+        for (ACCESSION_ENTITY element : accessionedElements) {
+            if (element.getVersion() >= maxVersion) {
+                maxVersion = element.getVersion();
+                lastVersionEntity = element;
+            }
+        }
+
+        return lastVersionEntity;
     }
 
     @Override
-    public void enableAccessions(List<AccessionWrapper<MODEL, String, ACCESSION>> accessionedObjects) {
-        customMethodsRepository.enableByHashedMessageIn(accessionedObjects.stream().map(AccessionWrapper::getHash)
-                .collect(Collectors.toSet()));
+    public ModelWrapper<MODEL, String, ACCESSION> findAccessionVersion(ACCESSION accession, int version)
+            throws AccessionDoesNotExistException, AccessionDeprecatedException, AccessionMergedException {
+        ACCESSION_ENTITY result = doFindAccessionVersion(accession, version);
+        return toModelWrapper(result);
+    }
+
+    private ACCESSION_ENTITY doFindAccessionVersion(ACCESSION accession, int version) throws
+            AccessionDoesNotExistException, AccessionMergedException, AccessionDeprecatedException {
+        ACCESSION_ENTITY result = repository.findByAccessionAndVersion(accession, version);
+        if (result == null) {
+            assertAccessionMergedOrDeprecated(accession);
+            throw new AccessionDoesNotExistException(accession.toString(), version);
+        }
+        return result;
     }
 
     @Override
-    public AccessionWrapper<MODEL, String, ACCESSION> update(AccessionWrapper<MODEL, String, ACCESSION> accession)
-            throws AccessionDoesNotExistException, HashAlreadyExistsException {
-        Collection<ENTITY> accessionedElements = repository.findByAccession(accession.getAccession());
-        assertAccessionExists(accession, accessionedElements);
-        assertHashDoesNotExist(accession);
+    public void insert(List<ModelWrapper<MODEL, String, ACCESSION>> objects) {
+        Set<ACCESSION_ENTITY> entitySet = objects.stream().map(toEntityFunction).collect(Collectors.toSet());
+        repository.insert(entitySet);
+    }
 
-        int version = 1;
-        for (ENTITY accessionedElement : accessionedElements) {
-            if (version <= accessionedElement.getVersion()) {
-                version = accessionedElement.getVersion() + 1;
+    @Override
+    public AccessionWrapper<MODEL, String, ACCESSION> patch(ModelWrapper<MODEL, String, ACCESSION> accession)
+            throws AccessionDoesNotExistException, HashAlreadyExistsException, AccessionDeprecatedException,
+            AccessionMergedException {
+        List<ACCESSION_ENTITY> entities = getAccession(accession.getAccession());
```

Comment:
I find this method's behaviour a bit strange, because it receives a wrapper but the version defined inside it is completely ignored. Could this cause any issues? 🤔

Reply:
It can't cause any issue, as the object is modified appropriatedly, the alternative is to modify the function in the services to use accession, hash and models as parameters so it is not really a big change. Do I proceed?

Reply:
done